crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::hash::{Hash, Hasher};
+
+use ahash::RandomState;
+
+use crate::bit_data::BitData;
+use crate::circuit_instruction::{
+    CircuitInstruction, ExtraInstructionAttributes, OperationFromPython,
+};
+use crate::dag_node::{DAGInNode, DAGNode, DAGOpNode, DAGOutNode};
+use crate::dot_utils::build_dot;
+use crate::error::DAGCircuitError;
+use crate::imports;
+use crate::interner::{IndexedInterner, Interner};
+use crate::operations::{Operation, OperationRef, Param, PyInstruction};
+use crate::packed_instruction::PackedInstruction;
+use crate::rustworkx_core_vnext::isomorphism;
+use crate::{BitType, Clbit, Qubit, TupleLikeArg};
+
+use hashbrown::{HashMap, HashSet};
+use indexmap::map::Entry;
+use indexmap::IndexMap;
+use itertools::Itertools;
+
+use pyo3::exceptions::{PyIndexError, PyRuntimeError, PyTypeError, PyValueError};
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{
+    IntoPyDict, PyDict, PyInt, PyIterator, PyList, PySequence, PySet, PyString, PyTuple, PyType,
+};
+
+use rustworkx_core::connectivity::connected_components as core_connected_components;
+use rustworkx_core::dag_algo::layers;
+use rustworkx_core::err::ContractError;
+use rustworkx_core::graph_ext::ContractNodesDirected;
+use rustworkx_core::petgraph;
+use rustworkx_core::petgraph::prelude::StableDiGraph;
+use rustworkx_core::petgraph::prelude::*;
+use rustworkx_core::petgraph::stable_graph::{EdgeReference, NodeIndex};
+use rustworkx_core::petgraph::unionfind::UnionFind;
+use rustworkx_core::petgraph::visit::{
+    EdgeIndexable, IntoEdgeReferences, IntoNodeReferences, NodeIndexable,
+};
+use rustworkx_core::petgraph::Incoming;
+use rustworkx_core::traversal::{
+    ancestors as core_ancestors, bfs_successors as core_bfs_successors,
+    descendants as core_descendants,
+};
+
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, VecDeque};
+use std::convert::Infallible;
+use std::f64::consts::PI;
+
+#[cfg(feature = "cache_pygates")]
+use std::cell::RefCell;
+
+static CONTROL_FLOW_OP_NAMES: [&str; 4] = ["for_loop", "while_loop", "if_else", "switch_case"];
+static SEMANTIC_EQ_SYMMETRIC: [&str; 4] = ["barrier", "swap", "break_loop", "continue_loop"];
+
+#[derive(Clone, Debug)]
+pub(crate) enum NodeType {
+    QubitIn(Qubit),
+    QubitOut(Qubit),
+    ClbitIn(Clbit),
+    ClbitOut(Clbit),
+    VarIn(PyObject),
+    VarOut(PyObject),
+    Operation(PackedInstruction),
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum Wire {
+    Qubit(Qubit),
+    Clbit(Clbit),
+    Var(PyObject),
+}
```

Comment:
Yeah, these definitely will need to be `pub` types. I did this in #12959 because I needed to match on `NodeType` in a few places.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/12550/commits/1d71204330125d9ee62103f537863af9281eb481

Reply:
The separate variants for combinations of wire type and in/out were actually from before `Wire` existed, but your suggestion is definitely what was intended, ultimately 🙂.

Reply:
Of course, if `pub` is needed rather than `pub(crate)` it will be forced on you when you try to use it.

It would be kind of nice if there were another visibility level because these functions are only meant to be used by a small set of crates.

Reply:
I suspect that I might end up inlining the `Wire` into the `NodeType` enum in a future PR that I have planned that'll better pack the `Wire` type for more efficient data use.

For the purposes of this PR, I think there's nothing more to do.